### PR TITLE
CDC library. Visibilty and richness/data access improvements

### PR DESCRIPTION
**Pull request by elcallio**
Split from #30 - the refactoring parts, mainly dealing with making certain types opaque interfaces, adding info/access to metadata and query objects etc.

### CODE REVIEW

**Comment by elcallio** [comment]
@avelanarius @haaawk - new, split PR.

File: scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java [hunk 1]
```diff
@@ -93,4 +95,30 @@ default boolean isDeleted(ChangeSchema.ColumnDefinition c) {
     default boolean getIsDeleted(String columnName) {
         return isDeleted(columnName);
     }
+
+    @Override
+    default Iterator<Cell> iterator() {
+        return stream().iterator();
```
Comment (haaawk) [inline scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java:110]: where does this `stream()` method come from/is defined? Same question to `dataStream()`.

File: scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java [hunk 2]
```diff
@@ -55,22 +56,7 @@ public ByteBuffer getUnsafeBytes(ColumnDefinition c) {
...
-                first = false;
-            }
-
-            String name = columnDefinition.getColumnName();
-            Cell cell = getCell(name);
-            result.append(name).append("=").append(cell);
-        }
-        result.append(")");
-        return result.toString();
+        return stream().map(Object::toString).collect(Collectors.joining(", ", "RawChange(", ")"));
```
Comment (haaawk) [inline scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java:59]: Why did you change Driver3RawChange( to RawChange(? It seems more correct to state the exact type, no?

File: scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java [hunk 3]
```diff
@@ -52,7 +54,7 @@ default Long getTTL() {
     default Object getAsObject(String columnName) {
         return getAsObject(getSchema().getColumnDefinition(columnName));
     }
-
+
```
Comment (haaawk) [inline scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java:68]: trailing whitespace

**Comment by haaawk** [comment]
I agree with most of the changes. My only concern is that we shouldn't expose raw bytes in all the versions of the library but only in those that enforce driver version. Maybe it's and overkill though since CQL protocol does not depend on the driver version.

File: scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java (hunk 2 above)
Comment (elcallio) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java:59]: No. The user does not know which implementation it is, and a driver4 impl should give the same output imho, If it as not for how inheritance works, it could be a defaulted method in the interface.

File: scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java (hunk 1 above)
Comment (elcallio) [reply scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java:110]: It is literally 11 lines below...

**Comment by elcallio** [comment]
> 
> 
> I agree with most of the changes. My only concern is that we shouldn't expose raw bytes in all the versions of the library but only in those that enforce driver version. Maybe it's and overkill though since CQL protocol does not depend on the driver version.

I am glad you opted to answer your own comment...

**Review by avelanarius — APPROVED** [review]
LGTM

File: scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java (hunk 1 above)
Comment (haaawk) [reply scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java:110]: Yes. I am literally blind sometimes :)

File: scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java (hunk 3 above)
Comment (haaawk) [reply scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java:68]: What about this one @elcallio ?

File: scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java (hunk 2 above)
Comment (haaawk) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java:59]: I don't think driver3 and driver4 have to return the same String. That's the purpose of overriding the method. toString is mostly used for logging/debugging so it's better to be more specific than not. toString is not an element of RawChange interface. It's an element of Object interface and its javadoc says:
```
Returns a string representation of the object.
In general, the {@code toString} method returns a string that "textually represents" this object.
```
It explicitly says **this** object not the superclass or anything else.

**Comment by haaawk** [comment]
> > I agree with most of the changes. My only concern is that we shouldn't expose raw bytes in all the versions of the library but only in those that enforce driver version. Maybe it's and overkill though since CQL protocol does not depend on the driver version.
> 
> I am glad you opted to answer your own comment...

Well. To be fair those bytes are useless/unsafe unless they are put into the driver connected using the same protocol version that the one used to obtain those bytes. I guess this isn't more unsafe than using raw bytes with a plain driver anyway so fine.

**Comment by haaawk** [comment]
@elcallio There's one trailing space left and a toString issue and we can merge as everything else looks good.

File: scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java (hunk 2 above)
Comment (elcallio) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java:59]: No, it says "this" object as in not another object. A toString should describe an object in a manner useful to the user. While not a pretty-print, it should be usable as a printf debugging tool. That means for data carrier interfaces like this, representing its contents, not its implementation type.

Comment (haaawk) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java:59]: I'm not fully convinced. Let's take an example from Swing:
superclass AbstractElement -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/javax/swing/text/AbstractDocument.java#L1764
one subclass BranchElement -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/javax/swing/text/AbstractDocument.java#L2302
another subclass LeftElement -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/javax/swing/text/AbstractDocument.java#L2519
they seem to be very specific in toString

File: scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java (hunk 3 above)
Comment (elcallio) [reply scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/RawChange.java]: Fine.

File: scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java (hunk 2 above)
Comment (elcallio) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java:59]: That is because their types carries important info - namely what type of node it is (leaf, branch etc...)

Comment (haaawk) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java:59]: So do Driver3RawChange and potential Driver4RawChange. Adding class name to toString seems to be a common practice. For example java.beans.PropertyChangeEvent will include the name of every subclass in its toString -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/java/beans/PropertyChangeEvent.java#L151

Comment (elcallio) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java]: Because it is the only aspect of naming in the hierarchy that carries meaning. In the case of Driver3RawChange, the one that carries meaning is instead RawChange. But I am to tired to debate. Now all printouts will carry pointless filler info and be harder to read.

Comment (haaawk) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java]: Thank you for making the codebase "worse" :)

Comment (elcallio) [reply scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java]: It is what I do. Wait, that did not sound right...
